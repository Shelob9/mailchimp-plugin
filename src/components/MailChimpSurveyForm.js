import React, {useState} from 'react';
import MailChimpForm from './MailChimpForm'
import PropTypes from 'prop-types';

function MailChimpSurveyForm(
	{
		emailField,
		initialQuestionId,
		questions,
		onChange,
		onBlur,
		findNextQuestion,
		findCurrentQuestion,
		listId,
	}
) {


	const [currentQuestionIndex, setCurrentQuestionIndex] = useState(
		questions.findIndex(question => initialQuestionId === question.fieldId)
	);


	const emailFieldId = emailField.fieldId;
	const submitButtonId = 'submitQuestion';
	const submitButton = {"fieldId": submitButtonId, "fieldType": "submit", "value": "Subscribe"};
	const initialQuestion = findCurrentQuestion(currentQuestionIndex, questions, currentQuestionIndex);

	const initialProcessor = {
		"type": "mc-subscribe",
		"listId": listId,
		"emailField": emailFieldId,
		"mergeFields": [],
		"groupFields": [initialQuestion.fieldId],
		"submitUrl": "https:\/\/formcalderas.lndo.site\/wp-json\/caldera-api\/v1\/messages\/mailchimp\/v1\/lists\/subscribe"
	};

	const [processor, setProcessor] = useState(initialProcessor);

	const initialForm = {
		fields: [
			initialQuestion,
			emailField,
			submitButton
		],
		rows: [
			{
				rowId: 'r1',
				columns: [
					{
						columnId: 'r1-c1',
						fields: [initialQuestion.fieldId],
						width: 1,
					}
				]
			},
			{
				rowId: 'r2',
				columns: [
					{
						columnId: 'r2-c1',
						fields: [emailFieldId],
						width: '1/2',
					},
					{
						columnId: 'r2-c2',
						fields: [submitButtonId],
						width: '1/2',
					}
				]
			}
		],
		processors: [processor]
	};

	const [form, setForm] = useState(initialForm);


	const updateForm = () => {
		const nextQuestion = (currentQuestionId, questions);
		if (nextQuestion) {
			setProcessor({
				...processor,
				groupFields: [initialQuestion.fieldId]
			});
			setForm({
				...form,
				processors: [
					...form.processors,
					processor
				]
			});
		} else {
			setProcessor({
				...processor,
				groupFields: []
			});
			setForm({
				...form,
				processors: [
					...form.processors,
					processor
				]
			});
		}


	};
	const onSubmit = (values) => {
		return new Promise((resolve, reject) => {
			onSubmit(values, processor).then(r => r.json()).then(r => {
					updateForm();
					resolve(new Response(JSON.stringify({message: 'next step!'})));

				})
				.catch(e => {
					reject(e);
				});


		});
	}


	return (
		<MailChimpForm
			form={form}
			onBlur={onBlur}
			onChange={onChange}
			onSubmit={onSubmit}
			hideOnSubmit={false}
		/>
	)
}


MailChimpSurveyForm.defaultProps = {

	onChange: () => {
	},
	onBlur: () => {
	},
	onSubmit: () => {
	},
	emailField: {
		"fieldId": "mc-email",
		"fieldType": "input",
		"html5Type": "email",
		"isRequired": true,
		"label": "Email",
		"default": ""
	},
	findNextFunction(currentQuestionIndex, questions) {
		if (-1 !== currentQuestionIndex && currentQuestionIndex + 1 <= questions.length) {
			return questions[currentQuestionIndex];
		}
		return false;
	},
	findCurrentQuestion(findIndex, questions, currentQuestionIndex) {
		if (-1 !== currentQuestionIndex) {
			return questions[findIndex];
		}
		return false;
	}
};

MailChimpSurveyForm.propTypes = {
	emailField: PropTypes.object.isRequired,
	listId: PropTypes.string.isRequired,
	initialQuestionId: PropTypes.string.isRequired,
	questions: PropTypes.array.isRequired,
	onChange: PropTypes.func,
	onBlur: PropTypes.func,
	onSubmit: PropTypes.func,
};

export default MailChimpSurveyForm;
