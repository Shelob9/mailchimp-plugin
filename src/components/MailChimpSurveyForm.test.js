import React from 'react';

import renderer from 'react-test-renderer';
import MailChimpSurveyForm from "./MailChimpSurveyForm";

const fakeQuestionFieldId2 = 'fakeGroup2';
const fakeQuestionField2 = {
	"fieldId": fakeQuestionFieldId2,
	"label": "Are you new to JavaScript",
	"fieldType": "checkbox",
	"value": [],
	"options": [
		{"id": "75091baaad", "value": "75091baaad", "label": "Totally"},
		{"id": "75091baaa1", "value": "75091baad1", "label": "Not At All"},
		{"id": "0a4dc88501", "value": "0a4dc88501", "label": "A Bit"}
	]
};

const fakeQuestionFieldId3 = 'fakeGroup3';
const fakeQuestionField3 = {
	"fieldId": fakeQuestionFieldId3,
	"label": "Are you comfortable with any JvaScr",
	"fieldType": "checkbox",
	"value": [],
	"options": [
		{"id": "15091baaad", "value": "15091baaad", "label": "React"},
		{"id": "15091baad1", "value": "15091baad1", "label": "VueJs"},
		{"id": "1a4dc88501", "value": "1a4dc88501", "label": "AngularJS"}
	]
};

const questions = [fakeQuestionField2, fakeQuestionField3];

const emailField = {
	type: 'input',
	html5type: 'email',
	required: true,
	label: 'Email',
}
it('Forms', () => {

	expect(renderer.create(<MailChimpSurveyForm
		listId={'a111'}
		onChange={() => {
		}}
		onBlur={() => {
		}}
		questions={questions}
		emailField={emailField}
		initialQuestionId={fakeQuestionFieldId2}
	/>)).toMatchSnapshot();


});


