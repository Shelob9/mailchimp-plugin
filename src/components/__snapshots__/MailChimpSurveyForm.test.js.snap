// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forms 1`] = `
Array [
  "",
  <div>
    
    <div
      className="caldera"
    >
      <form
        className="mc-subscribe-a111 caldera-form"
        onSubmit={[Function]}
      >
        <div
          className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
        >
          <div
            className="caldera-column width-1 sc-bdVaJa iHZvIS"
            id="r1-c1"
          >
            <div
              className="caldera-field-group caldera-field-area-checkbox caldera-field-wrapper caldera-field-wrapper-checkbox"
            >
              <input
                checked={false}
                className="caldera-field caldera-field-checkbox"
                id="fakeGroup2"
                onChange={[Function]}
                type="checkbox"
                value="1"
              />
              <label
                className="caldera-field-label caldera-field-label-checkbox"
                htmlFor="fakeGroup2"
              >
                Are you new to JavaScript
                 
              </label>
            </div>
          </div>
        </div>
        <div
          className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
        >
          <div
            className="caldera-column width-1 sc-bdVaJa iHZvIS"
            id="r2-c1"
          />
          <div
            className="caldera-column width-1 sc-bdVaJa iHZvIS"
            id="r2-c2"
          >
            <div
              className="caldera-field-group caldera-field-area-submit caldera-field-wrapper caldera-field-wrapper-submit"
            >
              <label
                className="caldera-field-label caldera-field-label-submit"
                htmlFor="submitQuestion"
              >
                
                 
              </label>
              <input
                className="caldera-field caldera-field-text"
                disabled={false}
                id="submitQuestion"
                type="submit"
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>,
]
`;
